src/shared/particle_dynamics/slender_structure_dynamics.h

```diff
@@ -0,0 +1,408 @@
+                F_[index_i].col(Dimensions - 1) = transformation_matrix_i * pseudo_n_[index_i];
+                F_[index_i].col(Dimensions - 2) = transformation_matrix_i * pseudo_b_n_[index_i];
+				F_bending_[index_i] = transformation_matrix_i * deformation_part_two * transformation_matrix_i.transpose() * B_[index_i];
+                F_b_bending_[index_i] = transformation_matrix_i * deformation_part_three * transformation_matrix_i.transpose() * B_[index_i];
+			};
+
+		protected:
+			StdLargeVec<Vecd> &pos_, &pseudo_n_, &n0_;
+			StdLargeVec<Matd> &B_, &F_, &F_bending_;
+			StdLargeVec<Matd> &transformation_matrix_;
+
+		    StdLargeVec<Vecd> &pseudo_b_n_, &b_n0_;
+            StdLargeVec<Matd> &F_b_bending_;
+		};
+
+		/**
+		 * @class BaseBarRelaxation
+		 * @brief abstract class for preparing bar relaxation
+		 */
+		class BaseBarRelaxation : public LocalDynamics, public BarDataInner
+		{
+		public:
+			explicit BaseBarRelaxation(BaseInnerRelation &inner_relation);
+			virtual ~BaseBarRelaxation(){};
+
+		protected:
+			StdLargeVec<Real> &rho_, &thickness_;
+			StdLargeVec<Vecd> &pos_, &vel_, &acc_, &acc_prior_;
+			StdLargeVec<Vecd> &n0_, &pseudo_n_, &dpseudo_n_dt_, &dpseudo_n_d2t_, &rotation_,
+				&angular_vel_, &dangular_vel_dt_;
+			StdLargeVec<Matd> &B_, &F_, &dF_dt_, &F_bending_, &dF_bending_dt_;
+			StdLargeVec<Matd> &transformation_matrix_;
+
+			StdLargeVec<Real> &width_;
+            StdLargeVec<Vecd> &b_n0_, &pseudo_b_n_, &dpseudo_b_n_dt_, &dpseudo_b_n_d2t_, &rotation_b_,
+                            &angular_b_vel_, dangular_b_vel_dt_;
+            StdLargeVec<Matd>  &F_b_bending_, &dF_b_bending_dt_;
+		};
+
+		/**
+		 * @class BarStressRelaxationFirstHalf
+		 * @brief computing stress relaxation process by Verlet time stepping
+		 * This is the first step
+		 */
+		class BarStressRelaxationFirstHalf : public BaseBarRelaxation
+		{
+		public:
+			explicit BarStressRelaxationFirstHalf(BaseInnerRelation &inner_relation,
+													int number_of_gaussian_points = 4, bool hourglass_control = false);
+			virtual ~BarStressRelaxationFirstHalf(){};
+			void initialization(size_t index_i, Real dt = 0.0);
+
+			inline void interaction(size_t index_i, Real dt = 0.0)
+			{
+				const Vecd &global_shear_stress_i = global_shear_stress_[index_i];
+				const Matd &global_stress_i = global_stress_[index_i];
+				const Matd &global_moment_i = global_moment_[index_i];
+                const Matd &global_b_moment_i = global_b_moment_[index_i];
+                const Vecd &global_b_shear_stress_i = global_b_shear_stress_[index_i];
+
+				Vecd acceleration = Vecd::Zero();
+				Vecd pseudo_normal_acceleration = global_shear_stress_i;
+                Vecd pseudo_b_normal_acceleration = global_b_shear_stress_i;
+
+				const Neighborhood &inner_neighborhood = inner_configuration_[index_i];
+				for (size_t n = 0; n != inner_neighborhood.current_size_; ++n)
+				{
+					size_t index_j = inner_neighborhood.j_[n];
+
+				if (hourglass_control_)
```

Comment:
Seems like the hourglass control algorithm and getCorrectedAlmansiStrin should be modified accordingly.